### PR TITLE
use NIST sessions wherever possible

### PR DESCRIPTION
**Pull request by maxtaco**
- NIST = **N**on-**I**nteractive **S**ession **T**oken
- all engine tests now work
- add a nist test
- sytests all work too:
  - TestRekey is aggressive with playing with the clock, so it can't work with NISTs
  - TestAccountDeadlock needs a csrf_token from sesscheck, so fix that on the server side

### CODE REVIEW

**Comment by maxtaco** [comment]
Needs https://github.com/keybase/keybase/pull/2351 to pass CI

File: go/engine/nist_test.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+	longTok3 := nist.Token().String()
+	require.True(t, len(longTok3) > 60, "should be a long token")
+	require.NotEqual(t, longTok, longTok3, "after expiration, should get a new token")
+
+	// As before, once it's successful, then, as above, we get a short NIST token.
+	nist.MarkSuccess()
+	nist, err = tc.G.ActiveDevice.NIST(ctx)
+	require.NoError(t, err, "no nist error")
+	tok = nist.Token().String()
+	require.True(t, len(tok) < 60, "should be a short token")
```
Comment (patrickxb) [inline go/engine/nist_test.go]: check to make sure the short token is different too?

File: go/libkb/api.go [hunk 2]
```diff
@@ -540,6 +570,22 @@ func (a *InternalAPIEngine) consumeHeaders(resp *http.Response) (err error) {
...
 	}
 
+	if nist != nil {
+		nistReply := resp.Header.Get("X-Keybase-Auth-NIST")
+		switch nistReply {
+		case "":
+			break
+		case "verified":
+			nist.MarkSuccess()
+			break
```
Comment (patrickxb) [inline go/libkb/api.go]: you can delete this (no fallthrough in go)

File: go/libkb/api.go [hunk 3]
```diff
@@ -540,6 +570,22 @@ func (a *InternalAPIEngine) consumeHeaders(resp *http.Response) (err error) {
...
+		case "failed":
+			nist.MarkFailure()
+			break
```
Comment (patrickxb) [inline go/libkb/api.go]: delete this

Comment (maxtaco) [reply go/libkb/api.go]: this = `break`?

File: go/engine/nist_test.go [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+	tc := SetupEngineTest(t, "nist")
+	defer tc.Cleanup()
+	fu := CreateAndSignupFakeUser(tc, "nst")
+
+	fakeClock := clockwork.NewFakeClockAt(time.Now())
+	tc.G.SetClock(fakeClock)
+	tc.G.ResetLoginState()
+
+	// Need to set active devices
+	Logout(tc)
```
Comment (patrickxb) [inline go/engine/nist_test.go]: how about adding

    nist, err := tc.G.ActiveDevice.NIST(ctx)
    require.Error(t, err, whatever)

to make sure you can't make one of these logged out.

File: go/libkb/api.go (hunk 3 above)
Comment (patrickxb) [reply go/libkb/api.go]: yes

**Comment by maxtaco** [comment]
Good idea. It's not an error right now, you just get a `nil` `NIST` back.  I think this should be OK...

**Comment by patrickxb** [comment]
That's fine, just as long as it isn't a token...

**Comment by maxtaco** [comment]
@mmaxim  are you OK with e35da37fdc528d82d8821aa4bb5672ef001dfbc8?

**Comment by maxtaco** [comment]
Merging over Jenkins, which is rekt.

**Pull request merged**